vtk-m@1.9.0 +cuda cuda_arch=90 fails build: CMake/VTKmDeviceAdapters.cmake:216 (message): Error detecting architecture flags for CUDA.  Please set VTKm_CUDA_Architecture manually.

### Steps to reproduce the issue

`vtk-m@1.9.0 +cuda cuda_arch=90` fails to build using:
* spack@develop (5c6c3b403bc4a8cd08b4e9303748933f3f79f5f7 from `Sun Jun 18 07:17:19 2023 -0400`)
* GCC 11.1.0
* CUDA 11.8.0
* Ubuntu 20.04, X86_64

Concrete spec: [vtk-m.spec.json.txt](https://github.com/spack/spack/files/11871666/vtk-m.spec.json.txt)

<details><summary>Concretization</summary>
<pre>
 -   vtk-m@1.9.0%gcc@11.1.0~64bitids+ascent_types+cuda+cuda_native+doubleprecision+examples+fpic~ipo~kokkos~logging+mpi+openmp+rendering~rocm+shared~tbb~testlib~virtuals build_system=cmake build_type=Release cuda_arch=90 generator=make arch=linux-ubuntu20.04-x86_64
[+]      ^cmake@3.26.3%gcc@11.1.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-ubuntu20.04-x86_64
[+]          ^ncurses@6.3%gcc@11.1.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]          ^openssl@1.1.1u%gcc@11.1.0~docs~shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-x86_64
[+]              ^ca-certificates-mozilla@2023-01-10%gcc@11.1.0 build_system=generic arch=linux-ubuntu20.04-x86_64
[+]              ^perl@5.36.0%gcc@11.1.0+cpanm+open+shared+threads build_system=generic arch=linux-ubuntu20.04-x86_64
[+]                  ^berkeley-db@18.1.40%gcc@11.1.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu20.04-x86_64
[+]                  ^bzip2@1.0.8%gcc@11.1.0~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-x86_64
[+]                  ^gdbm@1.23%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]              ^zlib@1.2.13%gcc@11.1.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-x86_64
[+]      ^cuda@11.8.0%gcc@11.1.0~allow-unsupported-compilers~dev build_system=generic arch=linux-ubuntu20.04-x86_64
[+]          ^libxml2@2.10.3%gcc@11.1.0~python build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]              ^libiconv@1.17%gcc@11.1.0 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]              ^xz@5.4.1%gcc@11.1.0+pic build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]      ^gmake@4.4.1%gcc@11.1.0~guile build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]      ^mpich@4.1.1%gcc@11.1.0~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~two_level_namespace~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu20.04-x86_64
[+]          ^findutils@4.9.0%gcc@11.1.0 build_system=autotools patches=440b954 arch=linux-ubuntu20.04-x86_64
[+]          ^libfabric@1.18.0%gcc@11.1.0~debug~kdreg build_system=autotools fabrics=rxm,sockets,tcp,udp arch=linux-ubuntu20.04-x86_64
[+]          ^libpciaccess@0.17%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]              ^libtool@2.4.7%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]              ^util-macros@1.19.3%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]          ^pkgconf@1.9.5%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]          ^yaksa@0.2%gcc@11.1.0~cuda~rocm build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]              ^autoconf@2.69%gcc@11.1.0 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-x86_64
[+]              ^automake@1.16.5%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]              ^m4@1.4.19%gcc@11.1.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-x86_64
[+]                  ^diffutils@3.9%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                  ^libsigsegv@2.14%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]              ^python@3.7.15%gcc@11.1.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,f2fd060 arch=linux-ubuntu20.04-x86_64
[+]                  ^expat@2.5.0%gcc@11.1.0+libbsd build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                      ^libbsd@0.11.7%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                          ^libmd@1.0.4%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                  ^gettext@0.21.1%gcc@11.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                      ^tar@1.34%gcc@11.1.0 build_system=autotools zip=pigz arch=linux-ubuntu20.04-x86_64
[+]                          ^pigz@2.7%gcc@11.1.0 build_system=makefile arch=linux-ubuntu20.04-x86_64
[+]                          ^zstd@1.5.5%gcc@11.1.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]                  ^libffi@3.4.4%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                  ^libxcrypt@4.4.33%gcc@11.1.0~obsolete_api build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                  ^readline@8.2%gcc@11.1.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-x86_64
[+]                  ^sqlite@3.42.0%gcc@11.1.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu20.04-x86_64
[+]                  ^util-linux-uuid@2.38.1%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
</pre>
</details>

Error:
```
$> spack install -f ./vtk-m.spec.json 
...
==> Installing vtk-m-1.9.0-okorbigkpgcosp6tambnnx5lamtd4n7e
==> No binary for vtk-m-1.9.0-okorbigkpgcosp6tambnnx5lamtd4n7e found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/12/12355dea1a24ec32767260068037adeb71abb3df2f9f920c92ce483f35ff46e4.tar.gz
==> No patches needed for vtk-m
==> vtk-m: Executing phase: 'cmake'
==> Error: ProcessError: Command exited with status 1:
    '/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/cmake-3.26.3-fok342qbpnpiznqxyzdvd4glkq22kvhu/bin/cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/vtk-m-1.9.0-okorbigkpgcosp6tambnnx5lamtd4n7e' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DBUILD_TESTING:BOOL=OFF' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=ON' '-DCMAKE_INSTALL_RPATH:STRING=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/vtk-m-1.9.0-okorbigkpgcosp6tambnnx5lamtd4n7e/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/vtk-m-1.9.0-okorbigkpgcosp6tambnnx5lamtd4n7e/lib64;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/libxml2-2.10.3-mxocfz7rp5lgy2yzsfqn7gf2yrrevzf7/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/libiconv-1.17-5f66p4aoivmlxa7htmpsffk5ov62vgci/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/xz-5.4.1-tppbn4xhprhizecunvu5egyztf7kmicv/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/zlib-1.2.13-hi2xsgcpyaect4ojyrhzeqj56vg7ldvj/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/mpich-4.1.1-pe7z6kruawbnt72svqy2bnsf2eurqc7j/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/libfabric-1.18.0-iv5kj6s4dciyn5f32scy3e24zxxjgzv3/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/libpciaccess-0.17-j5kbtglykqfohsexjzn5xgtkw27tf3kj/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/yaksa-0.2-o2bqqi7oirz77kdi65mm6gcvhqjpypru/lib;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/cuda-11.8.0-e5qmi3ynerocnxupx3epemw25cezxkoo/lib64' '-DCMAKE_PREFIX_PATH:STRING=/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/mpich-4.1.1-pe7z6kruawbnt72svqy2bnsf2eurqc7j;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/yaksa-0.2-o2bqqi7oirz77kdi65mm6gcvhqjpypru;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/libpciaccess-0.17-j5kbtglykqfohsexjzn5xgtkw27tf3kj;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/libfabric-1.18.0-iv5kj6s4dciyn5f32scy3e24zxxjgzv3;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/gmake-4.4.1-lp2ol6q2quzz7sjtfkrxeczlsvzukkzk;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/cuda-11.8.0-e5qmi3ynerocnxupx3epemw25cezxkoo;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/cuda-11.8.0-e5qmi3ynerocnxupx3epemw25cezxkoo/targets/x86_64-linux/lib/cmake;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/libxml2-2.10.3-mxocfz7rp5lgy2yzsfqn7gf2yrrevzf7;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/xz-5.4.1-tppbn4xhprhizecunvu5egyztf7kmicv;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/cmake-3.26.3-fok342qbpnpiznqxyzdvd4glkq22kvhu;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/zlib-1.2.13-hi2xsgcpyaect4ojyrhzeqj56vg7ldvj;/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/libiconv-1.17-5f66p4aoivmlxa7htmpsffk5ov62vgci' '-DVTKm_ENABLE_TESTING:BOOL=OFF' '-DBUILD_SHARED_LIBS=ON' '-DVTKm_USE_DOUBLE_PRECISION:BOOL=ON' '-DVTKm_ENABLE_LOGGING:BOOL=OFF' '-DVTKm_ENABLE_MPI:BOOL=ON' '-DVTKm_ENABLE_RENDERING:BOOL=ON' '-DVTKm_USE_DEFAULT_TYPES_FOR_ASCENT:BOOL=ON' '-DVTKm_NO_DEPRECATED_VIRTUAL:BOOL=ON' '-DVTKm_USE_64BIT_IDS:BOOL=OFF' '-DVTKm_ENABLE_CUDA:BOOL=ON' '-DCMAKE_CUDA_HOST_COMPILER=/usr/bin/g++' '-DVTKm_ENABLE_OPENMP:BOOL=ON' '-DVTKm_ENABLE_KOKKOS:BOOL=OFF' '-DVTKm_ENABLE_TBB:BOOL=OFF' '-DVTKm_INSTALL_EXAMPLES:BOOL=ON' '/tmp/root/spack-stage/spack-stage-vtk-m-1.9.0-okorbigkpgcosp6tambnnx5lamtd4n7e/spack-src'

1 error found in build log:
     17    -- The CUDA compiler identification is NVIDIA 11.8.89
     18    -- Detecting CUDA compiler ABI info
     19    -- Detecting CUDA compiler ABI info - done
     20    -- Check for working CUDA compiler: /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.1.0/cuda-11.8.0-e5qmi3ynerocnxupx3epemw25cezxkoo/bin/nvcc - skipped
     21    -- Detecting CUDA compile features
     22    -- Detecting CUDA compile features - done
  >> 23    CMake Error at CMake/VTKmDeviceAdapters.cmake:216 (message):
     24      Error detecting architecture flags for CUDA.  Please set
     25      VTKm_CUDA_Architecture manually.
     26    Call Stack (most recent call first):
     27      CMake/VTKmWrappers.cmake:14 (include)
     28      CMakeLists.txt:201 (include)
     29
```


### Error message

<details><summary>Error message</summary>
<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.21.0.dev0 (10fd13d9ba7d7ec301e62b95f0eca862aafadc91)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/11871678/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/11871679/spack-build-out.txt)

@kmorel @vicentebolea @kwryankrattiger @nchaimov @wspear

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/lib/spack/spack/build_systems/cmake.py
+++ b/lib/spack/spack/build_systems/cmake.py
@@ -296,7 +296,45 @@ class CMakeBuilder(BaseBuilder):
                 define("CMAKE_PREFIX_PATH", spack.build_environment.get_cmake_prefix_path(pkg)),
             ]
         )
+
         return args
+
+    @staticmethod
+    def define_cuda_architectures(pkg):
+        """Returns the str ``-DCMAKE_CUDA_ARCHITECTURES:STRING=(expanded cuda_arch)``.
+
+        ``cuda_arch`` is variant composed of a list of target CUDA architectures and
+        it is declared in the cuda package.
+
+        This method is no-op for cmake<3.18 and when ``cuda_arch`` variant is not set.
+
+        """
+        cmake_flag = str()
+        if "cuda_arch" in pkg.spec.variants and pkg.spec.satisfies("^cmake@3.18:"):
+            cmake_flag = CMakeBuilder.define(
+                "CMAKE_CUDA_ARCHITECTURES", pkg.spec.variants["cuda_arch"].value
+            )
+
+        return cmake_flag
+
+    @staticmethod
+    def define_hip_architectures(pkg):
+        """Returns the str ``-DCMAKE_HIP_ARCHITECTURES:STRING=(expanded amdgpu_target)``.
+
+        ``amdgpu_target`` is variant composed of a list of the target HIP
+        architectures and it is declared in the rocm package.
+
+        This method is no-op for cmake<3.18 and when ``amdgpu_target`` variant is
+        not set.
+
+        """
+        cmake_flag = str()
+        if "amdgpu_target" in pkg.spec.variants and pkg.spec.satisfies("^cmake@3.21:"):
+            cmake_flag = CMakeBuilder.define(
+                "CMAKE_HIP_ARCHITECTURES", pkg.spec.variants["amdgpu_target"].value
+            )
+
+        return cmake_flag
 
     @staticmethod
     def define(cmake_var, value):

--- a/lib/spack/spack/test/build_systems.py
+++ b/lib/spack/spack/test/build_systems.py
@@ -311,6 +311,16 @@ class TestCMakePackage(object):
         with pytest.raises(KeyError, match="not a variant"):
             s.package.define_from_variant("NONEXISTENT")
 
+    def test_cmake_std_args_cuda(self, default_mock_concretization):
+        s = default_mock_concretization("vtk-m +cuda cuda_arch=70 ^cmake@3.23")
+        option = spack.build_systems.cmake.CMakeBuilder.define_cuda_architectures(s.package)
+        assert "-DCMAKE_CUDA_ARCHITECTURES:STRING=70" == option
+
+    def test_cmake_std_args_hip(self, default_mock_concretization):
+        s = default_mock_concretization("vtk-m +rocm amdgpu_target=gfx900 ^cmake@3.23")
+        option = spack.build_systems.cmake.CMakeBuilder.define_hip_architectures(s.package)
+        assert "-DCMAKE_HIP_ARCHITECTURES:STRING=gfx900" == option
+
 
 @pytest.mark.usefixtures("config", "mock_packages")
 class TestDownloadMixins(object):

--- a/var/spack/repos/builtin.mock/packages/vtk-m/package.py
+++ b/var/spack/repos/builtin.mock/packages/vtk-m/package.py
@@ -1,0 +1,37 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class VtkM(CMakePackage):
+    """This is a fake vtk-m package used to demonstrate virtual package providers
+    with dependencies."""
+
+    homepage = "http://www.spack-fake-vtk-m.org"
+    url = "http://www.spack-fake-vtk-m.org/downloads/vtk-m-1.0.tar.gz"
+
+    version("1.0", md5="0123456789abcdef0123456789abcdef")
+
+    variant("cuda", default=False, description="Build with CUDA")
+    variant(
+        "cuda_arch",
+        description="CUDA architecture",
+        default="none",
+        values=("70", "none"),
+        multi=False,
+        when="+cuda",
+    )
+
+    variant("rocm", default=False, description="Enable ROCm support")
+    variant(
+        "amdgpu_target",
+        default="none",
+        description="AMD GPU architecture",
+        values=("gfx900", "none"),
+        multi=False,
+        when="+rocm",
+    )
+    depends_on("cmake@3.18:")

--- a/var/spack/repos/builtin/packages/vtk-m/package.py
+++ b/var/spack/repos/builtin/packages/vtk-m/package.py
@@ -135,6 +135,10 @@ class VtkM(CMakePackage, CudaPackage, ROCmPackage):
     # Patch
     patch("diy-include-cstddef.patch", when="@1.5.3:1.8.0")
 
+    # VTK-M PR#2972
+    # https://gitlab.kitware.com/vtk/vtk-m/-/merge_requests/2972
+    patch("vtkm-cuda-swap-conflict-pr2972.patch", when="@1.9 +cuda ^cuda@12:")
+
     def cmake_args(self):
         spec = self.spec
         options = []

--- a/var/spack/repos/builtin/packages/vtk-m/package.py
+++ b/var/spack/repos/builtin/packages/vtk-m/package.py
@@ -232,28 +232,28 @@ class VtkM(CMakePackage, CudaPackage, ROCmPackage):
                 options.append("-DVTKm_ENABLE_CUDA:BOOL=ON")
                 options.append("-DCMAKE_CUDA_HOST_COMPILER={0}".format(env["SPACK_CXX"]))
 
-                cuda_archs = spec.variants["cuda_arch"].value
-                if spec.satisfies("@1.9.0: ^cmake@3.18:"):
-                    str_cuda_archs = ";".join(str(e) for e in cuda_archs)
-                    options.append(f"-DCMAKE_CUDA_ARCHITECTURES={str_cuda_archs}")
-
-                # Only a CUDA Arch can be set with the legacy flag
-                elif cuda_archs[0] in gpu_name_table:
-                    vtkm_cuda_arch = gpu_name_table[cuda_archs[0]]
-                    options.append(f"-DVTKm_CUDA_Architecture={vtkm_cuda_arch}")
+                if spec.satisfies("@1.9.0:"):
+                    options.append(self.builder.define_cuda_architectures(self))
 
                 else:
-                    raise InstallError(
-                        f"cuda_arch={cuda_archs[0]} needs cmake>=3.18 & VTK-m>=1.9.0"
-                    )
+                    # VTKm_CUDA_Architecture only accepts a single CUDA arch
+                    num_cuda_arch = spec.variants["cuda_arch"].value[0]
+                    str_cuda_arch = str()
+
+                    try:
+                        str_cuda_arch = gpu_name_table[num_cuda_arch]
+                    except KeyError:
+                        raise InstallError(
+                            f"cuda_arch={num_cuda_arch} needs cmake>=3.18 & VTK-m>=1.9.0"
+                        )
+                    options.append(f"-DVTKm_CUDA_Architecture={str_cuda_arch}")
 
             else:
                 options.append("-DVTKm_ENABLE_CUDA:BOOL=OFF")
 
             # hip support
             if "+rocm" in spec:
-                archs = ",".join(self.spec.variants["amdgpu_target"].value)
-                options.append("-DCMAKE_HIP_ARCHITECTURES:STRING={0}".format(archs))
+                options.append(self.builder.define_hip_architectures(self))
 
             # openmp support
             if "+openmp" in spec:

--- a/var/spack/repos/builtin/packages/vtk-m/package.py
+++ b/var/spack/repos/builtin/packages/vtk-m/package.py
@@ -231,17 +231,22 @@ class VtkM(CMakePackage, CudaPackage, ROCmPackage):
             if "+cuda_native" in spec:
                 options.append("-DVTKm_ENABLE_CUDA:BOOL=ON")
                 options.append("-DCMAKE_CUDA_HOST_COMPILER={0}".format(env["SPACK_CXX"]))
-                if "cuda_arch" in spec.variants:
-                    cuda_value = spec.variants["cuda_arch"].value
-                    cuda_arch = cuda_value[0]
-                    if cuda_arch in gpu_name_table:
-                        vtkm_cuda_arch = gpu_name_table[cuda_arch]
-                        options.append("-DVTKm_CUDA_Architecture={0}".format(vtkm_cuda_arch))
+
+                cuda_archs = spec.variants["cuda_arch"].value
+                if spec.satisfies("@1.9.0: ^cmake@3.18:"):
+                    str_cuda_archs = ";".join(str(e) for e in cuda_archs)
+                    options.append(f"-DCMAKE_CUDA_ARCHITECTURES={str_cuda_archs}")
+
+                # Only a CUDA Arch can be set with the legacy flag
+                elif cuda_archs[0] in gpu_name_table:
+                    vtkm_cuda_arch = gpu_name_table[cuda_archs[0]]
+                    options.append(f"-DVTKm_CUDA_Architecture={vtkm_cuda_arch}")
+
                 else:
-                    # this fix is necessary if compiling platform has cuda, but
-                    # no devices (this is common for front end nodes on hpc
-                    # clusters). We choose volta as a lowest common denominator
-                    options.append("-DVTKm_CUDA_Architecture=volta")
+                    raise InstallError(
+                        f"cuda_arch={cuda_archs[0]} needs cmake>=3.18 & VTK-m>=1.9.0"
+                    )
+
             else:
                 options.append("-DVTKm_ENABLE_CUDA:BOOL=OFF")
 

--- a/var/spack/repos/builtin/packages/vtk-m/vtkm-cuda-swap-conflict-pr2972.patch
+++ b/var/spack/repos/builtin/packages/vtk-m/vtkm-cuda-swap-conflict-pr2972.patch
@@ -1,0 +1,63 @@
+diff -ruN spack-src/vtkm/exec/cuda/internal/ExecutionPolicy.h spack-src-patched/vtkm/exec/cuda/internal/ExecutionPolicy.h
+--- spack-src/vtkm/exec/cuda/internal/ExecutionPolicy.h	2022-10-11 12:07:59.000000000 -0700
++++ spack-src-patched/vtkm/exec/cuda/internal/ExecutionPolicy.h	2023-07-06 17:23:35.898388363 -0700
+@@ -17,6 +17,7 @@
+ #include <vtkm/exec/cuda/internal/ThrustPatches.h>
+ VTKM_THIRDPARTY_PRE_INCLUDE
+ #include <thrust/execution_policy.h>
++#include <thrust/sort.h>
+ #include <thrust/system/cuda/execution_policy.h>
+ #include <thrust/system/cuda/memory.h>
+ VTKM_THIRDPARTY_POST_INCLUDE
+diff -ruN spack-src/vtkm/Swap.h spack-src-patched/vtkm/Swap.h
+--- spack-src/vtkm/Swap.h	2022-10-11 12:07:59.000000000 -0700
++++ spack-src-patched/vtkm/Swap.h	2023-07-06 17:25:31.623393290 -0700
+@@ -24,21 +24,31 @@
+ 
+ /// Performs a swap operation. Safe to call from cuda code.
+ #if defined(VTKM_CUDA)
++// CUDA 12 adds a `cub::Swap` function that creates ambiguity with `vtkm::Swap`.
++// This happens when a function from the `cub` namespace is called with an object of a class
++// defined in the `vtkm` namespace as an argument. If that function has an unqualified call to
++// `Swap`, it results in ADL being used, causing the templated functions `cub::Swap` and
++// `vtkm::Swap` to conflict.
++#if defined(VTKM_CUDA_VERSION_MAJOR) && (VTKM_CUDA_VERSION_MAJOR >= 12) && \
++  defined(VTKM_CUDA_DEVICE_PASS)
++using cub::Swap;
++#else
+ template <typename T>
+-VTKM_EXEC_CONT void Swap(T& a, T& b)
++VTKM_EXEC_CONT inline void Swap(T& a, T& b)
+ {
+-  using namespace thrust;
++  using thrust::swap;
+   swap(a, b);
+ }
++#endif
+ #elif defined(VTKM_HIP)
+ template <typename T>
+-__host__ void Swap(T& a, T& b)
++__host__ inline void Swap(T& a, T& b)
+ {
+-  using namespace std;
++  using std::swap;
+   swap(a, b);
+ }
+ template <typename T>
+-__device__ void Swap(T& a, T& b)
++__device__ inline void Swap(T& a, T& b)
+ {
+   T temp = a;
+   a = b;
+@@ -46,9 +56,9 @@
+ }
+ #else
+ template <typename T>
+-VTKM_EXEC_CONT void Swap(T& a, T& b)
++VTKM_EXEC_CONT inline void Swap(T& a, T& b)
+ {
+-  using namespace std;
++  using std::swap;
+   swap(a, b);
+ }
+ #endif

